transports/quic/src/transport.rs

```diff
@@ -253,7 +254,7 @@ impl<P: Provider> Transport for GenTransport<P> {
         {
             return None;
         }
-        Some(observed.clone())
+        address_translation(listen, observed)
```

Comment:
I am not sure what issue this is addressing.

The `libp2p_core::address_translation` function takes the port of `listen` and the IP of `observed`. This is relevant in the case of TCP without port-reuse where one does not listen on the port of an outgoing connection, i.e. here were one does not listen of the port of `observed`.

The relevant section of the `address_translation` doc comment:

https://github.com/libp2p/rust-libp2p/blob/338e467927f8a0cba34eb24288f0be9b0972d05d/core/src/translation.rs#L29-L32

I argue that this is not relevant for QUIC, given that QUIC will use the same UDP port for both outgoing and incoming connections. The port of `observed` is either equal to the port of `listen`, or, if the local node is behind a NAT, equal to the public NAT port mapping of `listen`. In the former case `address_translation` is a no-op. In the latter case `address_translation` returns the wrong port, namely LAN port, not the public Internet port.

Reply:
Note that with the upcoming redesign of https://github.com/libp2p/rust-libp2p/pull/4568/ we can get rid of `Transport::address_translation`. Whether a connection has been established using port-reuse or not will be visible to a `NetworkBehaviour`. Thus e.g. `libp2p-identify` receiving an observed address can judge itself, whether it needs to alter any ports.

Reply:
Also see https://github.com/libp2p/rust-libp2p/pull/2289#discussion_r983428741 for past discussion.

Reply:
> I am not sure what issue this is addressing.

This addresses the problem of ephemeral ports with QUIC. At least when behind NAT with port forwarding, without address translation you'll end up with invalid external port. The more annoying thing is that it will be valid for the duration of Autonat probing, but when later anyone else tries to reach the node, the port is no longer functional.

> The port of observed is either equal to the port of listen, or, if the local node is behind a NAT, equal to the public NAT port mapping of listen.

This is most certainly not true in practice, at least in some cases. Even on my machine with pfSense acting as a router, I have seen random ports all over the place observed by remote nodes.

> In the former case address_translation is a no-op.

Yes

> In the latter case address_translation returns the wrong port, namely LAN port, not the public Internet port.

No, in this case address translation will result in correct port. I guess it depends on how to do port forwarding, it should be possible for router to map ports both ways, but I can't imagine many users actually doing it, it seems that external ports are mapped onto LAN ports, but probably rarely outgoing LAN ports are mapped onto specific WAN ports. This is my educated guess based on observation, I'm not claiming some consumer routers are not doing that.

> Note that with the upcoming redesign of https://github.com/libp2p/rust-libp2p/pull/4568 we can get rid of Transport::address_translation. Whether a connection has been established using port-reuse or not will be visible to a NetworkBehaviour. Thus e.g. libp2p-identify receiving an observed address can judge itself, whether it needs to alter any ports.

Interesting. Well, then this will need to be adjusted when that happens, but in the meantime network needs to function and lack of address translation was a big issue for us since many nodes were not able to discover their public addresses with QUIC (that we use almost exclusively in place of TCP now).

Reply:
> Also see [#2289 (comment)](https://github.com/libp2p/rust-libp2p/pull/2289#discussion_r983428741) for past discussion.

This means the current implementation is still wrong though as it will happily return an observed TCP address as the translated address. If QUIC is "before" TCP in the transport stack, it would return a wrong address, wouldn't it?

Reply:
See lines 252-256 above, the protocol is checked there and returns `None` if it doesn't match. Same is done in TCP and it works properly.

Reply:
> See lines 252-256 above, the protocol is checked there and returns `None` if it doesn't match. Same is done in TCP and it works properly.

Ah you are right, I missed that because the diff didn't show it :roll_eyes: 

I don't really understand why we need to change anything here then. QUIC doesn't have a concept of ephemeral ports so `address_translation` can't do anything useful. If you still observe changing ports with QUIC, it might be that you are sitting behind a symmetric NAT. We can't do anything against that I am afraid.

> The more annoying thing is that it will be valid for the duration of Autonat probing, but when later anyone else tries to reach the node, the port is no longer functional.

We are aware of this issue. Unfortunately, it is a design flaw in AutoNAT and one of the things that prompted the design of AutoNATv2. In v2, the server uses a different outgoing port to perform the dial-back, meaning we don't accidentally "hole-punch" through the NAT and get a more trustworthy result that the address is actually reachable.

Reply:
> If you still observe changing ports with QUIC, it might be that you are sitting behind a symmetric NAT. We can't do anything against that I am afraid.

But I still might have port forwarding configured such that the port matching my listening port is still reachable from the outside. This is what we recommend our users to do when they are behind NAT and it seems to work. It doesn't cover the case when forwarded port doesn't match listening port, in that case user has to specify external address explicitly, but it is still useful to do address translation, as mentioned above, in worst case it is no-op.